### PR TITLE
Update classifications

### PR DESCRIPTION
**Pull request by roshankern**
This PR is ready for review!

In this PR, classifications are also derived for the various models added to the `phenotypic_profiling` repo. This PR also adds READMEs describing the final files and their structure after processing in this repo is complete.

### CODE REVIEW

File: 4.classify-single-cell-phenotypes/4a.classify-single-cell-phenotypes/4a.classify-single-cell-phenotypes.py [hunk 1]
```diff
@@ -10,10 +10,9 @@
 
 
 import pathlib
-import urllib.request
 import joblib
```
Comment (axiomcura) [inline 4.classify-single-cell-phenotypes/4a.classify-single-cell-phenotypes/4a.classify-single-cell-phenotypes.py]: Consider moving joblib where pandas numpy numpy is to follow PEP8 import order. Joblib is not part of the python standard library. [Here](https://peps.python.org/pep-0008/#imports) is the documentation for import order!

File: 4.classify-single-cell-phenotypes/4a.classify-single-cell-phenotypes/4a.classify-single-cell-phenotypes.py [hunk 2]
```diff
@@ -69,17 +68,20 @@
     # load features
     col_types = {col: np.float32 for col in feature_cols}
     plate_features = pd.read_csv(
-        normalized_plate_path, low_memory=True, usecols=feature_cols
+        normalized_plate_path, low_memory=True, usecols=feature_cols,
```
Comment (axiomcura) [inline 4.classify-single-cell-phenotypes/4a.classify-single-cell-phenotypes/4a.classify-single-cell-phenotypes.py]: Looking at the `pandas` documentation for [`read_csv()`](https://pandas.pydata.org/docs/reference/api/pandas.read_csv.html), it appears that the `low_memory` parameter is set to `True` by default. Considering this default behavior, it might be worth considering whether to include this parameter unless there are specific reasons to explicitly state it.

File: 4.classify-single-cell-phenotypes/4a.classify-single-cell-phenotypes/4a.classify-single-cell-phenotypes.py [hunk 3]
```diff
@@ -69,17 +68,20 @@
...
     )
+    
     # load metadata
     print("Loading plate metadata...")
     col_types = {col: str for col in metadata_cols}
     plate_metadata = pd.read_csv(
-        normalized_plate_path, low_memory=True, usecols=metadata_cols
+        normalized_plate_path, low_memory=True, usecols=metadata_cols,
```
Comment (axiomcura) [inline 4.classify-single-cell-phenotypes/4a.classify-single-cell-phenotypes/4a.classify-single-cell-phenotypes.py]: Same comment as above

File: 4.classify-single-cell-phenotypes/4a.classify-single-cell-phenotypes/README.md [hunk 4]
```diff
@@ -0,0 +1,56 @@
+# 4a. Classify Single Cell Phenotypes
+
+In this module, we present our pipeline for classifying single cell phenotypes.
```
Comment (axiomcura) [inline 4.classify-single-cell-phenotypes/4a.classify-single-cell-phenotypes/README.md]:
```suggestion
In this module, we present our pipeline for classifying single-cell phenotypes.
```

File: 4.classify-single-cell-phenotypes/4a.classify-single-cell-phenotypes/README.md [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+
+## Deriving Classifications:
```
Comment (axiomcura) [inline 4.classify-single-cell-phenotypes/4a.classify-single-cell-phenotypes/README.md]:
```suggestion
## Deriving Classifications
```

File: 4.classify-single-cell-phenotypes/4a.classify-single-cell-phenotypes/README.md [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+
+Follow the steps in [4.classify-single-cell-phenotypes/README.md](../README.md) to run this module.
+
+## File Description
+
+This module outputs compressed csv files that contain the classification probabilities of preprocessed Cell Health features output by [3.preprocess-features](../../3.preprocess-features/).
+These classifications are derived per-plate by each of the models created in the [phenotypic_profiling](https://github.com/WayScience/phenotypic_profiling_model) repository.
+These models include multi-class models (16 output probabilities, 1 for each phenotypic class) and single-class models (2 output probabiltiies, positive and negative probabilities for the model's particular phenotypic class).
```
Comment (axiomcura) [inline 4.classify-single-cell-phenotypes/4a.classify-single-cell-phenotypes/README.md]:
```suggestion
These models include multi-class models (16 output probabilities, 1 for each phenotypic class) and single-class models (2 output probabilities, positive and negative probabilities for the model's particular phenotypic class).
```

File: 4.classify-single-cell-phenotypes/4a.classify-single-cell-phenotypes/README.md [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+
+In this module, we present our pipeline for classifying single cell phenotypes.
+
+## Deriving Classifications:
+
+Follow the steps in [4.classify-single-cell-phenotypes/README.md](../README.md) to run this module.
+
+## File Description
+
+This module outputs compressed csv files that contain the classification probabilities of preprocessed Cell Health features output by [3.preprocess-features](../../3.preprocess-features/).
```
Comment (axiomcura) [inline 4.classify-single-cell-phenotypes/4a.classify-single-cell-phenotypes/README.md]:
```suggestion
This module outputs compressed csv files containing the classification probabilities of preprocessed Cell-Health features output by [3.preprocess-features](../../3.preprocess-features/).
```

File: 4.classify-single-cell-phenotypes/4a.classify-single-cell-phenotypes/README.md [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+These classifications are derived per-plate by each of the models created in the [phenotypic_profiling](https://github.com/WayScience/phenotypic_profiling_model) repository.
```
Comment (axiomcura) [inline 4.classify-single-cell-phenotypes/4a.classify-single-cell-phenotypes/README.md]:
```suggestion
These classifications are derived per plate by each of the models created in the [phenotypic_profiling](https://github.com/WayScience/phenotypic_profiling_model) repository.
```

File: 4.classify-single-cell-phenotypes/4a.classify-single-cell-phenotypes/README.md [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+These models include multi-class models (16 output probabilities, 1 for each phenotypic class) and single-class models (2 output probabiltiies, positive and negative probabilities for the model's particular phenotypic class).
+For more information on model training and type see [phenotypic_profiling_model/2.train_model](https://github.com/WayScience/phenotypic_profiling_model/tree/main/2.train_model).
+
+Each compressed csv file output by this module contains rows of single-cell metadata (location, perturbation, etc) and feature classification probabilities, as derived by the particular model.
```
Comment (axiomcura) [inline 4.classify-single-cell-phenotypes/4a.classify-single-cell-phenotypes/README.md]:
```suggestion
Each compressed csv file generated by this module includes rows of single-cell metadata (location, perturbation, etc) along with feature classification probabilities, which is determined by the specific model used.
```
Dervied, for me, was a bit difficult to understand. Could be commonly used term the project. Feel free to reject!

File: 4.classify-single-cell-phenotypes/4a.classify-single-cell-phenotypes/README.md [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+There is one compressed csv file for each of the nine plates in the Cell Health dataset.
```
Comment (axiomcura) [inline 4.classify-single-cell-phenotypes/4a.classify-single-cell-phenotypes/README.md]:
```suggestion
There is one compressed csv file corresponding to each of the nine plates in the Cell-Health dataset.
```

File: 4.classify-single-cell-phenotypes/4a.classify-single-cell-phenotypes/README.md [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+| 1141.205          | 27.372            | 4             | G18           | SQ00014613     | SQ00014613_G18_04        | ARID1B-2         | 0.123      | 0.877               |
+| 896.493           | 42.108            | 4             | G18           | SQ00014613     | SQ00014613_G18_04        | ARID1B-2         | 0.194      | 0.806               |
+| 1571.713          | 30.771            | 4             | G18           | SQ00014613     | SQ00014613_G18_04        | ARID1B-2         | 0.191      | 0.809               |
+| ...          | ...            | ...             | ...           | ...     | ...        | ...         | ...      | ...               |
+| 1535.734          | 55.748            | 9             | P21           | SQ00014613     | SQ00014613_P21_09        | PSMA1-1         | 0.385      | 0.615               |
+
+## File Structure
+
+The output file structure of this module mirrors the structure of the models hosted at [phenotypic_profiling_model/2.train_model/models](https://github.com/WayScience/phenotypic_profiling_model/tree/main/2.train_model/models), with folders contianing plate classifications in place of the models.
```
Comment (axiomcura) [inline 4.classify-single-cell-phenotypes/4a.classify-single-cell-phenotypes/README.md]:
```suggestion
The output file structure of this module mirrors the structure of the models hosted at [phenotypic_profiling_model/2.train_model/models](https://github.com/WayScience/phenotypic_profiling_model/tree/main/2.train_model/models), with folders containing plate classifications in place of the models.
```

File: 4.classify-single-cell-phenotypes/4b.derive-classification-profiles/README.md [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+Follow the steps in [4.classify-single-cell-phenotypes/README.md](../README.md) to run this module.
+
+## File Description
+
+This module outputs tsv files that contain "classification profiles" for a particular model.
+These profiles are generated from the single-cell classification probabilities generated in [4a.classify-single-cell-phenotypes](../4a.classify-single-cell-phenotypes/).
+For each model, we find the mean of the single-cell classification probabilities across each perturbation and cell line to create a composite profile. 
+This aggregated data provides a summarized view of cell behavior for each perturbation/cell line combination, as predicted by each model.
+
+The contents of the tsv file containing the classification profiles for the `OutOfFocus` model predictions from are shown below:
```
Comment (axiomcura) [inline 4.classify-single-cell-phenotypes/4b.derive-classification-profiles/README.md]:
```suggestion
The contents of the tsv file containing the classification profiles for the `OutOfFocus` model predictions are shown below:
```

File: 4.classify-single-cell-phenotypes/4b.derive-classification-profiles/README.md [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+
+## Deriving Classification Profiles:
+
+Follow the steps in [4.classify-single-cell-phenotypes/README.md](../README.md) to run this module.
+
+## File Description
+
+This module outputs tsv files that contain "classification profiles" for a particular model.
+These profiles are generated from the single-cell classification probabilities generated in [4a.classify-single-cell-phenotypes](../4a.classify-single-cell-phenotypes/).
+For each model, we find the mean of the single-cell classification probabilities across each perturbation and cell line to create a composite profile.
```
Comment (axiomcura) [inline 4.classify-single-cell-phenotypes/4b.derive-classification-profiles/README.md]:
```suggestion
For each model, we find the mean of the single-cell classification probabilities across each perturbation and cell line to create a composite profile.
```

File: 4.classify-single-cell-phenotypes/classification_utils.py [hunk 14]
```diff
@@ -25,28 +25,30 @@ def get_probas_dataframe(
...
-    else:
-        cols_to_load = [col for col in plate_features.columns.to_list() if "P__" in col]
+    if "CP" in feature_type:
+        feature_cols = [col for col in all_cols if "CP__" in col]
+        if "zernike_only" in feature_type:
+            feature_cols = [col for col in feature_cols if "Zernike" in col]
+        if "areashape_only" in feature_type:
+            feature_cols = [col for col in feature_cols if "AreaShape" in col]
+        if "_and_DP" in feature_type:
+            feature_cols = [col for col in all_cols if "P__" in col]
```
Comment (axiomcura) [inline 4.classify-single-cell-phenotypes/classification_utils.py]:
```suggestion
        feature_cols = [col for col in all_cols if "CP__" in col]
        if "zernike_only" in feature_type:
            feature_cols = [col for col in feature_cols if "Zernike" in col]
        elif "areashape_only" in feature_type:
            feature_cols = [col for col in feature_cols if "AreaShape" in col]
        elif "_and_DP" in feature_type:
            feature_cols = [col for col in all_cols if "P__" in col]
        else:
            raise ValueError("Invalid feature type")
```
This question focuses more on the design of this block. Applying multiple `if` statements is exhaustive. Meaning, if the `feature_type` were to be found within the multiple `if` statements (e.g `zernike_only`) the subsequent `if` statements will still be executed. For example if there is a string `areashape_only_and_DP`, the `feature_cols` variable under the `areas_shape_only` will be overwritten by the `feature_cols` under the `_and_DP` `if` statement. A good way to avoid this is using `elif`. Where it will stop once it finds the condition. 

I do not know if this is the expected behavior! feel free to reject!

**Review by axiomcura — COMMENTED** [review]
LGTM! I've left some comments. What are your thoughts on having either @jenna-tomkinson  or @MattsonCam  approve? I'm also available to approve if needed

**Review by axiomcura — COMMENTED** [review]
LGTM! I've left some comments. What are your thoughts on having either Jenna or Cameron approve? I'm also available to approve if needed

File: 4.classify-single-cell-phenotypes/4a.classify-single-cell-phenotypes/README.md (hunk 9 above)
Comment (roshankern) [reply 4.classify-single-cell-phenotypes/4a.classify-single-cell-phenotypes/README.md]: Although I use derive throughout the project, I think this still adds clarity. Thanks!

File: 4.classify-single-cell-phenotypes/classification_utils.py (hunk 14 above)
Comment (roshankern) [reply 4.classify-single-cell-phenotypes/classification_utils.py]: I think using the `elif` here makes sense, thanks for the suggestion! I will incorporate on my end. We definitely do not want to raise an error if none are found because that is not expected behavior.

**Comment by roshankern** [comment]
I do think it would be worth @jenna-tomkinson and/or @MattsonCam reviewing and approving just to make sure they understand the PR and extracted file structure. Thanks @axiomcura!

**Comment by jenna-tomkinson** [comment]
@roshankern Looks like I can not review this PR unless requested. Can you add both Cameron and I as reviewers? Thanks!

**Comment by roshankern** [comment]
Forgot about that, sorry! You two have both been added.

File: 4.classify-single-cell-phenotypes/4a.classify-single-cell-phenotypes/4a.classify-single-cell-phenotypes.py
Comment (jenna-tomkinson) [inline 4.classify-single-cell-phenotypes/4a.classify-single-cell-phenotypes/4a.classify-single-cell-phenotypes.py:1]: Just a reminder, please make sure that this file has been nbconverted after running the notebook.

File: 4.classify-single-cell-phenotypes/4a.classify-single-cell-phenotypes/4a.classify-single-cell-phenotypes.py [hunk 15]
```diff
@@ -10,11 +10,10 @@
 
 
 import pathlib
-import urllib.request
-import joblib
-import importlib
 import sys
+import gc
```
Comment (jenna-tomkinson) [inline 4.classify-single-cell-phenotypes/4a.classify-single-cell-phenotypes/4a.classify-single-cell-phenotypes.py:14]: Recommend including a code comment regarding what `gc` here is used for.

File: 4.classify-single-cell-phenotypes/4a.classify-single-cell-phenotypes/README.md [hunk 16]
```diff
@@ -0,0 +1,56 @@
...
+
+In this module, we present our pipeline for classifying single-cell phenotypes.
+
+## Deriving Classifications
+
+Follow the steps in [4.classify-single-cell-phenotypes/README.md](../README.md) to run this module.
+
+## File Description
+
+This module outputs compressed csv files containing the classification probabilities of preprocessed Cell-Health features output by [3.preprocess-features](../../3.preprocess-features/).
```
Comment (jenna-tomkinson) [inline 4.classify-single-cell-phenotypes/4a.classify-single-cell-phenotypes/README.md]:
```suggestion
This module outputs compressed CSV files containing the classification probabilities of preprocessed Cell-Health features output by [3.preprocess-features](../../3.preprocess-features/).
```

File: 4.classify-single-cell-phenotypes/4a.classify-single-cell-phenotypes/README.md [hunk 17]
```diff
@@ -0,0 +1,56 @@
...
+These classifications are derived per plate by each of the models created in the [phenotypic_profiling](https://github.com/WayScience/phenotypic_profiling_model) repository.
+These models include multi-class models (16 output probabilities, 1 for each phenotypic class) and single-class models (2 output probabilities, positive and negative probabilities for the model's particular phenotypic class).
+For more information on model training and type see [phenotypic_profiling_model/2.train_model](https://github.com/WayScience/phenotypic_profiling_model/tree/main/2.train_model).
+
+Each compressed csv file generated by this module includes rows of single-cell metadata (location, perturbation, etc) along with feature classification probabilities, which is determined by the specific model used.
```
Comment (jenna-tomkinson) [inline 4.classify-single-cell-phenotypes/4a.classify-single-cell-phenotypes/README.md]:
```suggestion
Each compressed CSV file generated by this module includes rows of single-cell metadata (location, perturbation, etc) along with feature classification probabilities, which are determined by the specific model used.
```

File: 4.classify-single-cell-phenotypes/4a.classify-single-cell-phenotypes/4a.classify-single-cell-phenotypes.py [hunk 18]
```diff
@@ -69,17 +68,20 @@
...
     plate_features = pd.read_csv(
-        normalized_plate_path, low_memory=True, usecols=feature_cols
+        normalized_plate_path, usecols=feature_cols,
     )
+    
     # load metadata
     print("Loading plate metadata...")
     col_types = {col: str for col in metadata_cols}
     plate_metadata = pd.read_csv(
-        normalized_plate_path, low_memory=True, usecols=metadata_cols
```
Comment (MattsonCam) [inline 4.classify-single-cell-phenotypes/4a.classify-single-cell-phenotypes/4a.classify-single-cell-phenotypes.py:78]: Why did you remove the low_memory parameter here, just curious? Has `del` and `gc` worked better for you?

File: 4.classify-single-cell-phenotypes/4a.classify-single-cell-phenotypes/README.md [hunk 19]
```diff
@@ -0,0 +1,56 @@
...
+There is one compressed csv file corresponding to each of the nine plates in the Cell-Health dataset.
```
Comment (jenna-tomkinson) [inline 4.classify-single-cell-phenotypes/4a.classify-single-cell-phenotypes/README.md]:
```suggestion
There is one compressed CSV file corresponding to each of the nine plates in the Cell-Health dataset.
```

File: 4.classify-single-cell-phenotypes/4b.derive-classification-profiles/README.md [hunk 20]
```diff
@@ -0,0 +1,47 @@
...
+
+In this module, we present our pipeline for creating classification profiles.
+
+## Deriving Classification Profiles:
+
+Follow the steps in [4.classify-single-cell-phenotypes/README.md](../README.md) to run this module.
+
+## File Description
+
+This module outputs tsv files that contain "classification profiles" for a particular model.
```
Comment (jenna-tomkinson) [inline 4.classify-single-cell-phenotypes/4b.derive-classification-profiles/README.md]:
```suggestion
This module outputs TSV files that contain "classification profiles" for a particular model.
```

File: 4.classify-single-cell-phenotypes/4b.derive-classification-profiles/README.md [hunk 21]
```diff
@@ -0,0 +1,47 @@
...
+These profiles are generated from the single-cell classification probabilities generated in [4a.classify-single-cell-phenotypes](../4a.classify-single-cell-phenotypes/).
+For each model, we find the mean of the single-cell classification probabilities across each perturbation and cell line to create a composite profile.
+This aggregated data provides a summarized view of cell behavior for each perturbation/cell line combination, as predicted by each model.
+
+The contents of the tsv file containing the classification profiles for the `OutOfFocus` model predictions are shown below:
```
Comment (jenna-tomkinson) [inline 4.classify-single-cell-phenotypes/4b.derive-classification-profiles/README.md]:
```suggestion
The contents of the TSV file containing the classification profiles for the `OutOfFocus` model predictions are shown below:
```

File: 4.classify-single-cell-phenotypes/README.md [hunk 22]
```diff
@@ -72,3 +75,7 @@ Inside the notebook [4b.derive-classification-profiles.ipynb](4b.derive-classifi
 # Run this script to classify Cell Health features
 bash 4.classify-single-cell-phenotypes.sh
 ```
+
+## Ouputs
```
Comment (jenna-tomkinson) [inline 4.classify-single-cell-phenotypes/README.md]:
```suggestion
## Outputs
```

File: 4.classify-single-cell-phenotypes/4b.derive-classification-profiles/README.md (hunk 21 above)
Comment (jenna-tomkinson) [reply 4.classify-single-cell-phenotypes/4b.derive-classification-profiles/README.md]: Maybe recommend adding that this is an example of what an output looks like.

File: 4.classify-single-cell-phenotypes/4a.classify-single-cell-phenotypes/README.md [hunk 23]
```diff
@@ -0,0 +1,56 @@
...
+The contents of the compressed csv file containing `OutOfFocus` model predictions from plate `SQ00014613` are shown below:
```
Comment (jenna-tomkinson) [inline 4.classify-single-cell-phenotypes/4a.classify-single-cell-phenotypes/README.md]:
```suggestion
The contents of the compressed CSV file containing `OutOfFocus` model predictions from plate `SQ00014613` are shown below:
```

Comment (jenna-tomkinson) [reply 4.classify-single-cell-phenotypes/4a.classify-single-cell-phenotypes/README.md]: Might also add that this is an example of an output, but just a suggestion.

**Review by jenna-tomkinson — APPROVED** [review]
Nice PR! I added a few suggestions and small comments for you to address. Thank you @roshankern for running these models!

File: 4.classify-single-cell-phenotypes/4a.classify-single-cell-phenotypes/README.md [hunk 24]
```diff
@@ -0,0 +1,56 @@
...
+
+| Location_Center_X | Location_Center_Y | Metadata_Site | Metadata_Well | Metadata_Plate | Metadata_Plate_Map_Name | Metadata_Reagent | OutOfFocus | OutOfFocus Negative |
```
Comment (MattsonCam) [inline 4.classify-single-cell-phenotypes/4a.classify-single-cell-phenotypes/README.md:20]: Great documentation! Consider just including the table in a notebook, or reducing the number of example entries.

File: 4.classify-single-cell-phenotypes/4a.classify-single-cell-phenotypes/README.md [hunk 25]
```diff
@@ -0,0 +1,56 @@
...
+| 896.493           | 42.108            | 4             | G18           | SQ00014613     | SQ00014613_G18_04        | ARID1B-2         | 0.194      | 0.806               |
+| 1571.713          | 30.771            | 4             | G18           | SQ00014613     | SQ00014613_G18_04        | ARID1B-2         | 0.191      | 0.809               |
+| ...          | ...            | ...             | ...           | ...     | ...        | ...         | ...      | ...               |
+| 1535.734          | 55.748            | 9             | P21           | SQ00014613     | SQ00014613_P21_09        | PSMA1-1         | 0.385      | 0.615               |
+
+## File Structure
+
+The output file structure of this module mirrors the structure of the models hosted at [phenotypic_profiling_model/2.train_model/models](https://github.com/WayScience/phenotypic_profiling_model/tree/main/2.train_model/models), with folders containing plate classifications in place of the models.
+
+The output structure is as follows:
```
Comment (MattsonCam) [inline 4.classify-single-cell-phenotypes/4a.classify-single-cell-phenotypes/README.md:32]: Very comprehensive. This can be a good thing, but more information can also make it harder to follow or emphasize the most important info. Something to consider, but it's definitely a balance

File: 4.classify-single-cell-phenotypes/4b.derive-classification-profiles/README.md [hunk 26]
```diff
@@ -0,0 +1,47 @@
+# 4b. Derive Classification Profiles
```
Comment (MattsonCam) [inline 4.classify-single-cell-phenotypes/4b.derive-classification-profiles/README.md:1]: Again, very thorough!

**Review by MattsonCam — APPROVED** [review]
LGTM @roshankern, good job!

File: 4.classify-single-cell-phenotypes/4a.classify-single-cell-phenotypes/4a.classify-single-cell-phenotypes.py
Comment (roshankern) [reply 4.classify-single-cell-phenotypes/4a.classify-single-cell-phenotypes/4a.classify-single-cell-phenotypes.py:1]: Oh, I added a note about that in the `4a-classify-single-cell-phenotypes` README:

Note: The [4a.classify-single-cell-phenotypes.ipynb](https://github.com/roshankern/cell-health-data/blob/update-classifications/4.classify-single-cell-phenotypes/4a.classify-single-cell-phenotypes/4a.classify-single-cell-phenotypes.ipynb) and [4b.derive-classification-profiles.ipynb](https://github.com/roshankern/cell-health-data/blob/update-classifications/4.classify-single-cell-phenotypes/4b.derive-classification-profiles/4b.derive-classification-profiles.ipynb) were evaluated, but their respective python scripts were run to save memory. Thus, the outputs for these notebooks are incomplete.

Comment (jenna-tomkinson) [reply 4.classify-single-cell-phenotypes/4a.classify-single-cell-phenotypes/4a.classify-single-cell-phenotypes.py:1]: Ahh my bad! Maybe also including a note in the notebooks might help as well, but totally optional!

Comment (roshankern) [reply 4.classify-single-cell-phenotypes/4a.classify-single-cell-phenotypes/4a.classify-single-cell-phenotypes.py:1]: Done!

File: 4.classify-single-cell-phenotypes/4a.classify-single-cell-phenotypes/4a.classify-single-cell-phenotypes.py (hunk 18 above)
Comment (roshankern) [reply 4.classify-single-cell-phenotypes/4a.classify-single-cell-phenotypes/4a.classify-single-cell-phenotypes.py:78]: Great question! Although I did not realize it before, the `low_memory` argument is `True` by default ([docs](https://pandas.pydata.org/docs/reference/api/pandas.read_csv.html)). Thus, setting it here is useless.

In my experience, using `del` and `gc` can be successful in reducing memory usage. It was particularly useful here where I was just barely over the max memory threshold.
